### PR TITLE
Remove Object Async. With Args.

### PR DESCRIPTION
**Pull request by BigUstad**
Also Functional test for ListObjects with Versions with RemoveObject/RemoveBucket

### CODE REVIEW

File: Minio/DataModel/ObjectOperationsArgs.cs [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+                if (versionIDs == null || versionIDs.Count < 1)
+                {
+                    throw new MinioException("AllVersions object set but List of version IDs is not set.");
+                }
+                this.AllVersions = true;
+                WithAddObjectNameWithVersions(rmArgs.ObjectName, versionIDs);
+            }
+        }
+
+        public RemoveMultipleObjectArgs WithAddObjectName(string name)
```
Comment (kannappanr) [inline Minio/DataModel/ObjectOperationsArgs.cs:93]:
```suggestion
        public RemoveMultipleObjectArgs WithObjectName(string name)
```

File: Minio/ApiEndpoints/ObjectOperations.cs [hunk 2]
```diff
@@ -36,6 +36,53 @@ public partial class MinioClient : IObjectOperations
...
+            {
+                request = await this.CreateRequest(args).ConfigureAwait(false);
+            }
+            else if (args.AllVersions)
+            {
+                string prefix = null;
+                string obj = args.ObjectName;
+                if (args.ObjectName.Contains("/"))
+                {
+                    prefix = args.ObjectName.Substring(0,args.ObjectName.IndexOf('/'));
```
Comment (kannappanr) [inline Minio/ApiEndpoints/ObjectOperations.cs:64]: what if there are multiple `/`

File: Minio/DataModel/ObjectOperationsArgs.cs (hunk 1 above)
Comment (BigUstad) [reply Minio/DataModel/ObjectOperationsArgs.cs:93]: WithBucketName/WithObjectName are already inherited from BucketArgs/ObjectArgs. I'll modify the name.

File: Minio/ApiEndpoints/ObjectOperations.cs (hunk 2 above)
Comment (BigUstad) [reply Minio/ApiEndpoints/ObjectOperations.cs]: Replaced.

File: Minio/DataModel/ObjectOperationsArgs.cs (hunk 1 above)
Comment (BigUstad) [reply Minio/DataModel/ObjectOperationsArgs.cs]: Changed to `WithSingleObject`

**Review by poornas — CHANGES_REQUESTED** [review]

File: Minio/DataModel/ObjectOperationsArgs.cs [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        internal List<string> ObjectNames { get; private set; }
+        // Each element in the list is a Tuple. Each Tuple has an Object name(string)
+        // And the second element is 
+        internal List<Tuple<string, List<string>>> ObjectNamesVersions  { get; private set; }
+
+        public RemoveObjectsArgs()
+        {
+            this.ObjectName = null;
+            this.ObjectNames = new List<string>();
+            this.ObjectNamesVersions = new List<Tuple<string, List<string>>>();
```
[the remark this answers is not in the document]
Comment (BigUstad) [reply Minio/DataModel/ObjectOperationsArgs.cs:59]: For  List<Tuple<string,string>>. My thinking was if there is one single object & multiple versions, then the user will have to enter the object name repeatedly. So, I used  List<Tuple<string, List<string>>>() instead. Tuple<object_name, List<multiple_version_ids>>. I can change it to List<Tuple<string,string>>.

[the remark this answers is not in the document]
Comment (poornas) [reply Minio/DataModel/ObjectOperationsArgs.cs:59]: you can still provide a .WithVersionIDs() method for the RemoveObjectAsync api call - and internally convert the args to a List<Tuple<string,string>> so that user doesn't have to repeat the objectName.  Otherwise the implementation looks needlessly complicated to batch the calls if a List<Tuple<string, List>> is used

File: Minio.Functional.Tests/FunctionalTest.cs [hunk 4]
```diff
@@ -470,6 +470,14 @@ internal async static Task TearDown(MinioClient minio, string bucketName)
             await minio.RemoveBucketAsync(rbArgs);
         }
 
+        internal async static Task RemoveObjects(MinioClient minio, string bucketName)
```
[the remark this answers is not in the document]
Comment (BigUstad) [reply Minio.Functional.Tests/FunctionalTest.cs:473]: Yes, change implemented.

File: Minio.Functional.Tests/FunctionalTest.cs [hunk 5]
```diff
@@ -2078,24 +2219,75 @@ internal async static Task ListObjectVersions_Test1(MinioClient minio)
...
+                    {
+                        size = filestream.Length;
+                    }
 
-                await minio.RemoveObjectAsync(bucketName, objectName + "0");
-                await minio.RemoveObjectAsync(bucketName, objectName + "1");
-                await minio.RemoveObjectAsync(bucketName, objectName + "2");
-                await minio.RemoveObjectAsync(bucketName, objectName + "3");
-                await TearDown(minio, bucketName);
+                    for(int i=0; i < 4; ++i)
```
[the remark this answers is not in the document]
Comment (BigUstad) [reply Minio.Functional.Tests/FunctionalTest.cs:2240]: Yes, changed.

File: Minio/Helper/OperationsHelper.cs [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+                    {
+                        listIndex = 0;
+                        fullErrorsList.AddRange(await callRemoveObjectVersions(args, objVersions, fullErrorsList, cancellationToken));
+                        objVersions.Clear();
+                    }
+                }
+                // Once we make the iteration count for remove as 1000.
+                // Check what is remaining in the tuple list.
+                // Remaining is more than 1000.
+                else if ((listIndex + objVerTuple.Item2.Count) > 1000)
```
[the remark this answers is not in the document]
Comment (BigUstad) [reply Minio/Helper/OperationsHelper.cs]: It got changed after shifting to List<Tuple<string, string>>.

File: Minio/DataModel/ObjectOperationsArgs.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (BigUstad) [reply Minio/DataModel/ObjectOperationsArgs.cs]: Changed to this Tuple.

File: Docs/API.md [hunk 7]
```diff
@@ -1507,14 +1524,52 @@ __Example__
...
+        {
+            Console.WriteLine("Listed all delete errors for remove objects on  " + bucketName + "\n");
+        });
+}
+catch (MinioException e)
+{
+    Console.WriteLine("Error: " + e);
+}
+
+// 2. Remove list of objects' (only specific versions mentioned in Version ID list) from the bucket bucketName
```
[the remark this answers is not in the document]
Comment (BigUstad) [reply Docs/API.md]: Changed.

File: Minio/ApiEndpoints/ObjectOperations.cs [hunk 8]
```diff
@@ -906,46 +920,15 @@ private async Task<List<DeleteError>> removeObjectsAsync(string bucketName, List
         /// <returns></returns>
         public async Task<IObservable<DeleteError>> RemoveObjectAsync(string bucketName, IEnumerable<string> objectNames, CancellationToken cancellationToken = default(CancellationToken))
         {
-            if (objectNames == null)
```
Comment (kannappanr) [inline Minio/ApiEndpoints/ObjectOperations.cs:909]: what will happen if `objectNames == null` ? Shouldn't this code be not removed?

File: Minio/DataModel/BucketOperationsArgs.cs [hunk 9]
```diff
@@ -141,7 +141,7 @@ public override RestRequest BuildRequest(RestRequest request)
         {
             request.AddQueryParameter("delimiter",this.Delimiter);
             request.AddQueryParameter("prefix",this.Prefix);
-            request.AddQueryParameter("max-keys", "1000");
+            //request.AddQueryParameter("max-keys", "1000");
```
Comment (kannappanr) [inline Minio/DataModel/BucketOperationsArgs.cs:144]: why was this code commented?

File: Minio/Minio.csproj [hunk 10]
```diff
@@ -34,14 +34,14 @@
     <PackageReference Include="Crc32.NET" Version="1.2.0"/>
     <PackageReference Include="SourceLink.Embed.AllSourceFiles" Version="2.8.3" PrivateAssets="All"/>
     <PackageReference Include="AsyncFixer" Version="1.1.6" PrivateAssets="All"/>
-    <PackageReference Include="RestSharp" Version="106.10.1"/>
+    <PackageReference Include="RestSharp" Version="106.11.7"/>
```
Comment (kannappanr) [inline Minio/Minio.csproj:37]: why is this change needed?

File: Minio/MinioClient.cs [hunk 11]
```diff
@@ -222,6 +222,7 @@ private void ArgsCheck(Args args)
                 utils.ValidateObjectName(objectName);
             }
 
+
```
Comment (kannappanr) [inline Minio/MinioClient.cs:225]:
```suggestion
```

File: Docs/API.md [hunk 12]
```diff
@@ -1873,31 +1873,49 @@ __Example__
...
 ```
 <a name="removeObjects"></a>
-### RemoveObjectAsync(string bucketName, IEnumerable<string> objectsList)
+### RemoveObjectsAsync(RemoveObjectsArgs args)
 
-`Task<IObservable<DeleteError>> RemoveObjectAsync(string bucketName, IEnumerable<string> objectsList, CancellationToken cancellationToken = default(CancellationToken))`
+`Task<IObservable<DeleteError>> RemoveObjectsAsync(RemoveObjectsArgs args, CancellationToken cancellationToken = default(CancellationToken))`
 
-Removes a list of objects.
+Removes a list of objects or a list of specified versionIDs of each object in the list.
```
Comment (poornas) [inline Docs/API.md:1911]:
```suggestion
Removes a list of objects or object versions.
```

File: Docs/API.md [hunk 13]
```diff
@@ -1856,8 +1857,7 @@ __Parameters__
 
 |Param   | Type	  | Description  |
 |:--- |:--- |:--- |
-| ``bucketName``  | _string_  | Name of the bucket  |
-| ``objectName``  | _string_  | Object name in the bucket |
+| ``args``  | _RemoveObjectArgs_ | Arguments Object - name.  |
```
Comment (poornas) [inline Docs/API.md:1860]:
```suggestion
| ``args``  | _RemoveObjectArgs_ | Arguments Object  |
```

File: Docs/API.md [hunk 14]
```diff
@@ -1913,14 +1931,52 @@ __Example__
...
+    RemoveObjectsAsync rmArgs = new RemoveObjectsAsync()
+                                            .WithBucket(bucketName)
+                                            .WithObjects(objectNames);
+    IObservable<DeleteError> observable = await minio.RemoveObjectAsync(rmArgs);
+    IDisposable subscription = observable.Subscribe(
+        deleteError => Console.WriteLine("Object: {0}", deleteError.Key),
+        ex => Console.WriteLine("OnError: {0}", ex),
+        () =>
+        {
+            Console.WriteLine("Listed all delete errors for remove objects on  " + bucketName + "\n");
```
Comment (poornas) [inline Docs/API.md:1951]:
```suggestion
            Console.WriteLine("Removed objects from " + bucketName + "\n");
```

File: Minio.Examples/Cases/RemoveObjects.cs [hunk 15]
```diff
@@ -17,32 +17,53 @@
...
+                    RemoveObjectsArgs objArgs = new RemoveObjectsArgs()
+                                                        .WithBucket(bucketName)
+                                                        .WithObjects(objectsList);
+                    IObservable<DeleteError> objectsOservable = await minio.RemoveObjectsAsync(objArgs).ConfigureAwait(false);
+                    IDisposable objectsSubscription = objectsOservable.Subscribe(
+                        objDeleteError => Console.WriteLine($"Object: {objDeleteError.Key}"),
+                            ex => Console.WriteLine($"OnError: {ex}"),
+                            () =>
+                            {
+                                Console.WriteLine($"Listed all delete errors for remove objects on {bucketName}\n");
```
Comment (poornas) [inline Minio.Examples/Cases/RemoveObjects.cs:46]:
```suggestion
                                Console.WriteLine($"Removed objects in list from {bucketName}\n");
```

File: Minio.Examples/Cases/RemoveObjects.cs [hunk 16]
```diff
@@ -17,32 +17,53 @@
...
-                   () =>
-                   {
-                       Console.WriteLine($"Listed all delete errors for remove objects on {bucketName}\n");
-                   });
+                    objVerDeleteError => Console.WriteLine($"Object: {objVerDeleteError.Key} " +
+                        $"Object Version: {objVerDeleteError.VersionId}"),
+                        ex => Console.WriteLine($"OnError: {ex}"),
+                        () =>
+                        {
+                            Console.WriteLine($"Listed all delete errors for remove objects with versions on {bucketName}\n");
```
Comment (poornas) [inline Minio.Examples/Cases/RemoveObjects.cs:60]:
```suggestion
                            Console.WriteLine($"Removed objects versions from {bucketName}\n");
```

File: Minio/ApiEndpoints/ObjectOperations.cs [hunk 17]
```diff
@@ -331,6 +331,58 @@ public async Task<Tagging> GetObjectTagsAsync(GetObjectTagsArgs args, Cancellati
...
+        /// <param name="cancellationToken">Optional cancellation token to cancel the operation</param>
+        /// <returns>Task</returns>
+        public async Task RemoveObjectAsync(RemoveObjectArgs args, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            args.Validate();
+            RestRequest request =  null;
+            if (string.IsNullOrEmpty(args.VersionId))
+            {
+                request = await this.CreateRequest(Method.DELETE, args.BucketName, args.ObjectName).ConfigureAwait(false);
+            }
```
Comment (poornas) [inline Minio/ApiEndpoints/ObjectOperations.cs:347]: can this not be handled by this.CreateRequest(args)

File: Minio/DataModel/ObjectOperationsArgs.cs [hunk 18]
```diff
@@ -533,6 +534,130 @@ public GetObjectArgs WithFile(string file)
...
+            return this;
+        }
+    }
+
+    public class RemoveObjectsArgs : ObjectArgs<RemoveObjectsArgs>
+    {
+        internal List<string> ObjectNames { get; private set; }
+        // Each element in the list is a Tuple. Each Tuple has an Object name(string)
+        // And the second element is 
+        //internal List<Tuple<string, List<string>>> ObjectNamesVersions  { get; private set; }
```
Comment (poornas) [inline Minio/DataModel/ObjectOperationsArgs.cs:566]: can be removed

**Review by poornas — CHANGES_REQUESTED** [review]

File: Minio/ApiEndpoints/ObjectOperations.cs (hunk 8 above)
Comment (BigUstad) [reply Minio/ApiEndpoints/ObjectOperations.cs:1032]: This function calls the new implementation.

File: Minio.Functional.Tests/FunctionalTest.cs [hunk 19]
```diff
@@ -2462,6 +2760,73 @@ internal async static Task RemoveObjects_Test2(MinioClient minio)
...
+            try
+            {
+                int count = 50;
+                Task[] tasks = new Task[count];
+                List<string> objectsList = new List<string>();
+                await Setup_WithLock_Test(minio, bucketName);
+                for (int i = 0; i < count; i++)
+                {
+                    tasks[i] = PutObject_Task(minio, bucketName, objectName + i.ToString(), null, null, 0, null, rsg.GenerateStreamFromSeed(5));
+                    tasks[i] = PutObject_Task(minio, bucketName, objectName + i.ToString(), null, null, 0, null, rsg.GenerateStreamFromSeed(5));
```
[the remark this answers is not in the document]
Comment (BigUstad) [reply Minio.Functional.Tests/FunctionalTest.cs]: Changed the loop.

File: Minio/ApiEndpoints/ObjectOperations.cs [hunk 20]
```diff
@@ -331,6 +331,58 @@ public async Task<Tagging> GetObjectTagsAsync(GetObjectTagsArgs args, Cancellati
...
+        /// <summary>
+        /// Removes multiple objects from a specific bucket
+        /// </summary>
+        /// <param name="args">RemoveObjectsArgs Arguments Object encapsulates information like - bucket name, List of objects, optional list of versions (for each object) to be deleted</param>
+        /// <param name="cancellationToken">Optional cancellation token to cancel the operation</param>
+        /// <returns>Observable that returns delete error while deleting objects if any</returns>
+        public async Task<IObservable<DeleteError>> RemoveObjectsAsync(RemoveObjectsArgs args, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            args.Validate();
+            List<DeleteError> fullErrorsList = new List<DeleteError>();
```
[the remark this answers is not in the document]
Comment (BigUstad) [reply Minio/ApiEndpoints/ObjectOperations.cs]: Changed.

File: Docs/API.md (hunk 12 above)
Comment (BigUstad) [reply Docs/API.md]: Changed.

File: Docs/API.md (hunk 13 above)
Comment (BigUstad) [reply Docs/API.md]: Changed.

File: Docs/API.md (hunk 14 above)
Comment (BigUstad) [reply Docs/API.md]: Changed message.

File: Minio.Examples/Cases/RemoveObjects.cs (hunk 15 above)
Comment (BigUstad) [reply Minio.Examples/Cases/RemoveObjects.cs]: Changed message.

File: Minio.Examples/Cases/RemoveObjects.cs (hunk 16 above)
Comment (BigUstad) [reply Minio.Examples/Cases/RemoveObjects.cs]: Changed message.

File: Minio/ApiEndpoints/ObjectOperations.cs (hunk 17 above)
Comment (BigUstad) [reply Minio/ApiEndpoints/ObjectOperations.cs]: Corrected the previous oversight. Also tested.

File: Minio/DataModel/ObjectOperationsArgs.cs (hunk 18 above)
Comment (BigUstad) [reply Minio/DataModel/ObjectOperationsArgs.cs]: Commented code removed.

**Review by kannappanr — APPROVED** [review]
LGTM

**Review by poornas — APPROVED** [review]
LGTM
